### ISSUE TITLE
`:not()` does not seem to be adding prefix

### ISSUE THREAD

**Issue by cgood92** [zeit/styled-jsx#146]
I have the following line:

```
.root > :global(*):not(header) {
...
}
```

Which get's rendered (after the babel transform) to 

```
.root[data-jsx=\"2663302060\"] > *:not(header) {
...
}
```

I guess I would expect the stuff inside the "NOT" to have prefixes as well, unless I stated that they were global.  So, what is being done and I what I would expect are different.

Here is what I would expect:

```
/* with prefix */
.root > :global(*):not(header)
=> 
.root[data-jsx=\"2663302060\"] > *:not(header[data-jsx=\"2663302060\"])
```

```
/* with global */
.root > :global(*):not(:global(header))
=> 
.root[data-jsx=\"2663302060\"] > *:not(header)
```

Styled-jsx version 0.5.7, as per dependency of next version 2.0.0-beta.32

**Comment by thysultan** [comment]
Maybe [L57](https://github.com/zeit/styled-jsx/blob/master/lib/style-transform.js#L57) can also `.replace(/&/g, prefix)` to allow for & in :global() i.e 

```scss
// will allow this pattern
.root > :global(*):not(header) {}

// to be written as
.root > :global(*:not(&)) {}
```

**Comment by giuseppeg** [comment]
I think that we shouldn't introduce the `&` symbol. 

Regexp are not good to match parenthesis therefore I think that we should do regular parsing like in my rewrite wip branch.

https://git.io/vyzFg
https://git.io/vyzFr
https://git.io/vyzF6

**Comment by thysultan** [comment]
@giuseppeg re: parenthesis, I think i found a bulletproof solution [https://regex101.com/r/5ssplS/1](https://regex101.com/r/5ssplS/1) `:global\(((?:\[.*\]|[^()]*|\([^()]*\))*)\)`

**Comment by giuseppeg** [comment]
@thysultan we could use a module like this https://github.com/juliangruber/balanced-match

**Comment by thysultan** [comment]
@giuseppeg I'm using regex

```js
/:global\(((?:[^\(\)\[\]]*|\[.*\]|\([^\(\)]*\))*)\)/g
```

in stylis, it hasn't failed any of the tests i have put it through.

**Comment by giuseppeg** [comment]
related #158

**Comment by giuseppeg** [comment]
@thysultan any chance that you can help us out fixing this one?

Here are some test cases that should pass:

```css
p:hover {
  color: red;
}

p::before {
  color: red;
}

:hover {
  color: red;
}

::before {
  color: red;
}

:hover p {
  color: red;
}

:global(a + b :hover, ::before a:not(b)) a {
  color: red;
}

a:not( a +b foo:hover, :global(marquee) a) > :hover {
  color: red;
}

.root > :global(*):not(header) {
  color: red;
}
`

**Comment by thysultan** [comment]
@giuseppeg What's the expected output?

**Comment by giuseppeg** [comment]
```css
p[data-jsx="abc"]:hover {
  color: red;
}

p[data-jsx="abc"]::before {
  color: red;
}

[data-jsx="abc"]:hover {
  color: red;
}

[data-jsx="abc"]::before {
  color: red;
}

[data-jsx="abc"]:hover p[data-jsx="abc"] {
  color: red;
}

a + b :hover ::before a:not(b) a[data-jsx="abc"] {
  color: red;
}

a[data-jsx="abc"]:not( a[data-jsx="abc"] +b[data-jsx="abc"] foo[data-jsx="abc"]:hover marquee a[data-jsx="abc"]) > [data-jsx="abc"]:hover {
  color: red;
}

.root[data-jsx="abc"] > *:not(header[data-jsx="abc"]) {
  color: red;
}
```

**Comment by giuseppeg** [comment]
@thysultan also we should fix #158 
Would it make sense to use [postcss-selector-parser](https://github.com/postcss/postcss-selector-parser) and simplify all the things? Regexp are very unstable

**Comment by thysultan** [comment]
I have a working version that passes the tests in [V3](https://github.com/thysultan/stylis.js/blob/master/rewrite/stylis.js?ts=2#L452-L509) will update the stylis dependency here when i release it, just working on supporting nesting for the cascade isolation mode. ex.

```
h1 {
   div {
      color: red;
   }
}

// expected output
h1#h4sh div#h4sh {color: red;}

// current output
h1#h4sh divh1#h4sh {color: red;}
```

**Comment by giuseppeg** [comment]
ok sounds good, just for my information what's the ETA?

**Comment by thysultan** [comment]
Want to release this weekend, will see how it goes.

**Comment by thysultan** [comment]
@giuseppeg These are the tests i have for covering cascade isolation [spec.js#L576-L674](https://github.com/thysultan/stylis.js/blob/master/rewrite/tests/spec.js#L576-L674)..

Where there any other edge cases?

**Comment by giuseppeg** [comment]
@thysultan oh cool! 
I guess any deep nested parentheses so if you can immagine some edge cases that'd be great. I just put those together a bit randomly. But otherwise the known issues are:

* the one from this issue
* https://github.com/zeit/styled-jsx/pull/168#issuecomment-292735128
* https://github.com/zeit/styled-jsx/issues/158 and probably other combinations like `:global(.foo #bar > baz)`

**Comment by thysultan** [comment]
@giuseppeg [V3](https://github.com/thysultan/stylis.js/blob/master/CHANGELOG.md) is out. After the update i guess the only parts that will remain in `lib/transform` are the source maps bits.

**Comment by giuseppeg** [comment]
@thysultan you rock man! thank you a lot for the great work :)

**Comment by giuseppeg** [comment]
Fixed in #210
